### PR TITLE
ssl proxy module support ntls.

### PR DESCRIPTION
**Pull request by wa5i**
rt

### CODE REVIEW

**Comment by wa5i** [comment]
@dongbeiouba

**Review by InfoHunter — APPROVED** [review]

File: modules/ngx_openssl_ntls/t/ntls_proxy.t [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+use lib 'lib';
+use Test::Nginx;
+use lib ".";
+use CA qw/ make_sm2_ca_subca_end_certs make_rsa_end_cert make_ec_end_cert /;
+###############################################################################
+
+select STDERR; $| = 1;
+select STDOUT; $| = 1;
+
+my $openssl = $ENV{'TEST_OPENSSL_BINARY'} || "/opt/babassl/bin/openssl";
```
Comment (dongbeiouba) [inline modules/ngx_openssl_ntls/t/ntls_proxy.t:23]: babassl -> tongsuo

File: modules/ngx_openssl_ntls/t/ntls_proxy.t [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        location / {
+            proxy_enable_ntls $arg_enable_ntls;
+            proxy_ssl_sign_certificate        client_sign.crt;
+            proxy_ssl_sign_certificate_key    client_sign.key;
+            proxy_ssl_enc_certificate         client_enc.crt;
+            proxy_ssl_enc_certificate_key     client_enc.key;
+            proxy_ssl_trusted_certificate     server_ca_chain.crt;
+            proxy_ssl_ciphers "ECC-SM2-WITH-SM4-SM3:ECDHE-SM2-WITH-SM4-SM3:RSA";
+
+			proxy_pass https://localhost:8082;
```
Comment (dongbeiouba) [inline modules/ngx_openssl_ntls/t/ntls_proxy.t:91]: 没对齐

File: modules/ngx_openssl_ntls/t/ntls_proxy.t [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        location /ecdhe {
+            proxy_enable_ntls $arg_enable_ntls;
+            proxy_ssl_sign_certificate        client_sign.crt;
+            proxy_ssl_sign_certificate_key    client_sign.key;
+            proxy_ssl_enc_certificate         client_enc.crt;
+            proxy_ssl_enc_certificate_key     client_enc.key;
+            proxy_ssl_trusted_certificate     server_ca_chain.crt;
+            proxy_ssl_ciphers "ECDHE-SM2-SM4-GCM-SM3:RSA";
+
+			proxy_pass https://localhost:8082;
```
Comment (dongbeiouba) [inline modules/ngx_openssl_ntls/t/ntls_proxy.t:103]: 没对齐

File: modules/ngx_openssl_ntls/t/stream_ntls_proxy.t [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+use lib 'lib';
+use Test::Nginx;
+use lib ".";
+use CA qw/ make_sm2_ca_subca_end_certs make_rsa_end_cert make_ec_end_cert /;
+###############################################################################
+
+select STDERR; $| = 1;
+select STDOUT; $| = 1;
+
+my $openssl = $ENV{'TEST_OPENSSL_BINARY'} || "/opt/babassl/bin/openssl";
```
Comment (dongbeiouba) [inline modules/ngx_openssl_ntls/t/stream_ntls_proxy.t:23]: babassl -> tongsuo

File: src/stream/ngx_stream_proxy_module.c [hunk 5]
```diff
@@ -2080,6 +2147,18 @@ ngx_stream_proxy_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
...
+    ngx_conf_merge_value(conf->enable_ntls, prev->enable_ntls, 0);
+    ngx_conf_merge_str_value(conf->enc_certificate,
+                             prev->enc_certificate, "");
+    ngx_conf_merge_str_value(conf->enc_certificate_key,
+                             prev->enc_certificate_key, "");
+    ngx_conf_merge_str_value(conf->sign_certificate,
+                             prev->sign_certificate, "");
+    ngx_conf_merge_str_value(conf->sign_certificate_key,
+                             prev->sign_certificate_key, "");
+    //conf->upstream.ssl_ciphers = conf->ssl_ciphers;
```
Comment (dongbeiouba) [inline src/stream/ngx_stream_proxy_module.c:2160]: 没用就删除吧

File: modules/ngx_openssl_ntls/t/ntls_proxy.t [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+my $ret7 = `echo -e "GET /ecdhe HTTP/1.0\r\n\r\n" | $openssl s_client -connect localhost:8083 -cipher ECDHE-SM2-SM4-CBC-SM3 -quiet -ign_eof -enc_cert $d/client_enc.crt -enc_key $d/client_enc.key -sign_cert $d/client_sign.crt -sign_key $d/client_sign.key -enable_ntls -ntls 2>&1`;
+my $ret8 = `echo -e "GET /ecdhe?enable_ntls=on HTTP/1.0\r\n\r\n" | $openssl s_client -connect localhost:8083 -cipher ECDHE-SM2-SM4-GCM-SM3 -quiet -ign_eof -enc_cert $d/client_enc.crt -enc_key $d/client_enc.key -sign_cert $d/client_sign.crt -sign_key $d/client_sign.key -enable_ntls -ntls 2>&1`;
+my $ret9 = `echo -e "GET /ecdhe?enable_ntls=off HTTP/1.0\r\n\r\n" | $openssl s_client -connect localhost:8083 -cipher ECDHE-SM2-SM4-GCM-SM3 -quiet -ign_eof -enc_cert $d/client_enc.crt -enc_key $d/client_enc.key -sign_cert $d/client_sign.crt -sign_key $d/client_sign.key -enable_ntls -ntls 2>&1`;
+
+like($ret1, qr/^ssl_protocal=TLSv1\.2.*$/m, 'client -----(TLSv1.2 AES128-GCM-SHA256)-----> server(no proxy_enable_ntls) -----(TLSv1.2)-----> origin');
+like($ret2, qr/^ssl_protocal=TLSv1\.2.*$/m, 'client -----(NTLSv1.1 ECC-SM2-SM4-GCM-SM3)-----> server(no proxy_enable_ntls) -----(TLSv1.2)-----> origin');
```
Comment (dongbeiouba) [inline modules/ngx_openssl_ntls/t/ntls_proxy.t:131]: ret2是CBC，不是GCM

**Review by dongbeiouba — CHANGES_REQUESTED** [review]

File: src/http/modules/ngx_http_grpc_module.c [hunk 7]
```diff
@@ -435,6 +442,42 @@ static ngx_command_t  ngx_http_grpc_commands[] = {
...
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_grpc_loc_conf_t, enc_certificate),
+      NULL },
+
+    { ngx_string("grpc_ssl_enc_certificate_key"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_grpc_loc_conf_t, enc_certificate_key),
+      NULL },
```
Comment (chobits) [inline src/http/modules/ngx_http_grpc_module.c:465]: a newline should be appended under L465

File: src/http/modules/ngx_http_proxy_module.c [hunk 8]
```diff
@@ -731,6 +738,41 @@ static ngx_command_t  ngx_http_proxy_commands[] = {
...
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_proxy_loc_conf_t, enc_certificate),
+      NULL },
+
+    { ngx_string("proxy_ssl_enc_certificate_key"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_proxy_loc_conf_t, enc_certificate_key),
+      NULL },
```
Comment (chobits) [inline src/http/modules/ngx_http_proxy_module.c:761]: newline need to be appended

File: src/http/modules/ngx_http_uwsgi_module.c [hunk 9]
```diff
@@ -569,6 +576,41 @@ static ngx_command_t ngx_http_uwsgi_commands[] = {
...
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_uwsgi_loc_conf_t, enc_certificate),
+      NULL },
+
+    { ngx_string("uwsgi_ssl_enc_certificate_key"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_uwsgi_loc_conf_t, enc_certificate_key),
+      NULL },
```
Comment (chobits) [inline src/http/modules/ngx_http_uwsgi_module.c:599]: newline

File: src/stream/ngx_stream_proxy_module.c [hunk 10]
```diff
@@ -336,6 +345,41 @@ static ngx_command_t  ngx_stream_proxy_commands[] = {
...
+      NGX_STREAM_SRV_CONF_OFFSET,
+      offsetof(ngx_stream_proxy_srv_conf_t, enc_certificate),
+      NULL },
+
+    { ngx_string("proxy_ssl_enc_certificate_key"),
+      NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_STREAM_SRV_CONF_OFFSET,
+      offsetof(ngx_stream_proxy_srv_conf_t, enc_certificate_key),
+      NULL },
```
Comment (chobits) [inline src/stream/ngx_stream_proxy_module.c:368]: \n

File: modules/ngx_openssl_ntls/t/ntls_proxy.t (hunk 1 above)
Comment (wa5i) [reply modules/ngx_openssl_ntls/t/ntls_proxy.t]: > babassl -> tongsuo

fixed

File: modules/ngx_openssl_ntls/t/ntls_proxy.t (hunk 3 above)
Comment (wa5i) [reply modules/ngx_openssl_ntls/t/ntls_proxy.t]: > 没对齐

fixed

File: src/stream/ngx_stream_proxy_module.c (hunk 5 above)
Comment (wa5i) [reply src/stream/ngx_stream_proxy_module.c]: > 没用就删除吧

fixed

File: src/http/modules/ngx_http_grpc_module.c (hunk 7 above)
Comment (wa5i) [reply src/http/modules/ngx_http_grpc_module.c:465]: > a newline should be appended under L465

fixed

File: modules/ngx_openssl_ntls/t/ntls_proxy.t (hunk 6 above)
Comment (wa5i) [reply modules/ngx_openssl_ntls/t/ntls_proxy.t]: > ret2是CBC，不是GCM

fixed

File: src/http/modules/ngx_http_proxy_module.c (hunk 8 above)
Comment (wa5i) [reply src/http/modules/ngx_http_proxy_module.c:761]: > newline need to be appended

fixed

File: src/stream/ngx_stream_proxy_module.c (hunk 10 above)
Comment (wa5i) [reply src/stream/ngx_stream_proxy_module.c:368]: > \n

fixed

**Pull request merged**